frame/merkle-mountain-range/src/lib.rs

```diff
@@ -0,0 +1,190 @@
+	/// any additional [LeafData](primitives::LeafDataProvider) defined by this type.
+	type LeafData: primitives::LeafDataProvider;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as MerkleMountainRange {
+		/// Latest MMR Root hash.
+		pub RootHash get(fn mmr_root_hash): <T as Trait>::Hash;
+
+		/// Current size of the MMR (number of leaves).
+		pub NumberOfLeaves get(fn mmr_leaves): u64;
+
+		/// Hashes of the nodes in the MMR.
+		///
+		/// Note this collection only contains MMR peaks, the inner nodes (and leaves)
+		/// are pruned and only stored in the Offchain DB.
+		pub Nodes get(fn mmr_peak): map hasher(identity) u64 => Option<<T as Trait>::Hash>;
+	}
+}
+
+decl_module! {
+	/// A public part of the pallet.
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			use primitives::LeafDataProvider;
+			let leaves = Self::mmr_leaves();
+			let peaks_before = mmr::utils::NodesUtils::new(leaves).number_of_peaks();
+			let (data, leaf_weight) = T::LeafData::leaf_data();
+			// append new leaf to MMR
+			let mut mmr: ModuleMMR<mmr::storage::RuntimeStorage, T> = mmr::MMR::new(leaves);
+			mmr.push(data).expect("MMR push never fails.");
+
+			// update the size
+			let (leaves, root) = mmr.finalize().expect("MMR finalize never fails.");
+			<NumberOfLeaves>::put(leaves);
+			<RootHash<T>>::put(root);
+
+			let peaks_after = mmr::utils::NodesUtils::new(leaves).number_of_peaks();
+
+			leaf_weight + <T as frame_system::Trait>::DbWeight::get().reads_writes(
+				2 + peaks_before,
+				2 + peaks_after,
+			)
+		}
+	}
+}
+
+/// A MMR specific to the pallet.
+type ModuleMMR<StorageType, T> = mmr::MMR<StorageType, T, LeafOf<T>>;
+
+/// Leaf data.
+type LeafOf<T> = <<T as Trait>::LeafData as primitives::LeafDataProvider>::LeafData;
+
+/// Hashing used for the pallet.
+pub(crate) type HashingOf<T> = <T as crate::Trait>::Hashing;
+
+impl<T: Trait> Module<T> {
+	fn offchain_key(pos: u64) -> Vec<u8> {
+		(T::INDEXING_PREFIX, pos).encode()
+	}
+
+	/// Generate a MMR proof for given `leaf_index`.
+	///
+	/// Note this method can only be used from an off-chain context
+	/// (Offchain Worker or Runtime API call), since it requires
+	/// all the leaves to be present.
+	/// It may return an error or panic if used incorrectly.
+	pub fn generate_proof(leaf_index: u64) -> Result<
+		(LeafOf<T>, primitives::Proof<<T as Trait>::Hash>),
+		mmr::Error,
+	> {
+		let mmr: ModuleMMR<mmr::storage::OffchainStorage, T> = mmr::MMR::new(Self::mmr_leaves());
+		mmr.generate_proof(leaf_index)
+	}
+
+	/// Verify MMR proof for given `leaf`.
+	///
+	/// This method is safe to use within the runtime code.
+	/// It will return `Ok(())` if the proof is valid
+	/// and an `Err(..)` if MMR is inconsistent (some leaves are missing)
+	/// or the proof is invalid.
+	pub fn verify_leaf(leaf: LeafOf<T>, proof: primitives::Proof<<T as Trait>::Hash>) -> Result<(), mmr::Error> {
```

Comment:
Probably dumb q, but what is the point of having `fn verify_leaf()` in the same pallet that is generating proofs? :) Is it going to prove something to itself?

If you're going to verify proofs coming from other MMRs, then you can't use nodes (hashes) from the `RuntimeStorage`, because `RuntimeStorage` contains nodes generated by this MMR.

Reply:
- `verify_proof` is usable on-chain by other pallets if needed (requires only `RuntimeStorage`)
- `generate_proof` is planned to be exposed via `RuntimeAPI` since it requires off-chain storage access, I believe it's still required to have the method `pub` though.

Reply:
I should have asked with more details - sorry, I'm still have this dumb q :)

Imagine that the instance `I1` of this pallet is deployed on chain `C1`. Let it builds `MMR1`. We may call `generate_proof()` on the instance `I1`, we will get `MMR1`-based proof `P1`. If I understand correctly, verifying this proof `P1` requires `MMR1.Peaks`. And the runtime `I1.RuntimeStorage` is supposed to store these peaks.

Now imagine other instance `I2` deployed on `C2`. It builds `MMR2` and the runtime storage `I2.RuntimeStorage` is storing peaks of `MMR2`.

So the question is: whether we want `P1` to be verified by `I1`, or by `I2`? Or both? My guess is that proof is assumed to be verified by other instance of the pallet. So `P1` should be verified by calling `I2::verify_leaf()`. If that is true, then, I believe, it won't work, because:
1) this function [is directly using](https://github.com/paritytech/substrate/pull/7312/files#diff-95496f2048e0ac8c7eb99001a4eef8ce939cadcddbda0611b4fc0fa4aa5011c0R191) `I2.RuntimeStorage`;
2) `I2.RuntimeStorage` is storing `MMR2.Peaks`, not the `MMR1.Peaks`;
3) we need `MMR1.Peaks` to verify proof `P1`.

I'm probably missing something, just trying to clarify here :)

Reply:
I see, it's not a dumb question at all :)

So the pallet is not symmetrical. This one's main purpose is to generate MMR root hashes from arbitrary on-chain data. The MMR root hash is going to be used by [BEEFY](https://github.com/paritytech/grandpa-bridge-gadget) and signed by the validator set. Conceptually BEEFY generates a stream of signed MMR root hashes (+ authority set handoffs).

This stream can then be consumed by a BEEFY Light Client (non-existent pallet yet). Instead of importing headers like we do in a regular light client, we import the MMR root hash (at least one per epoch, but can more). So `C2` will need BEEFY Light Client of `C1` deployed, so that it will know the last finalized MMR root hash (`MMR1.Root`; peaks are not needed for anything, they are just internal representation).
If `P1` is done against the root hash that is known by `C2` it can be verified there. If we want a bi-directional bridge, we would need to deploy BEEFY Light Client of `C2` on `C1` as well, so that `MMR2.Root` can be known on `C1` and `P2` can be verified there.

`I1::verify_leaf` is also useful on `C1` internally (outside of bridging context). Currently `frame_system` pallet stores up to `BlockHashCount` recent block hashes - the old ones are pruned. With MMR you will be able to prove on-chain at any point in time what the past block hash was. Obviously this applies to any historical data that might sporadically be required on-chain (for instance past validator set for equivocation reporting, etc).

Hope it clear things up. Perhaps I should add this to the high-level documentation of the pallet (will do tomorrow).

Reply:
Ok, thanks! I guess using it to verify historical data makes sense.